File: python/problems/leetcode/0124_binary_tree_maximum_path_sum/solution.py
"""LeetCode 124: Binary Tree Maximum Path Sum."""

from __future__ import annotations

from math import inf
from typing import Optional


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: Optional["TreeNode"] = None,
        right: Optional["TreeNode"] = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        best = -inf

        def dfs(node: Optional[TreeNode]) -> int:
            nonlocal best
            if node is None:
                return 0
            l = max(dfs(node.left), 0)
            r = max(dfs(node.right), 0)
            best = max(best, node.val + l + r)
            return node.val + max(l, r)

        dfs(root)
        return int(best)


if __name__ == "__main__":
    assert Solution().maxPathSum(TreeNode(1, TreeNode(2), TreeNode(3))) == 6
    root = TreeNode(-10, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    assert Solution().maxPathSum(root) == 42
    assert Solution().maxPathSum(TreeNode(-3)) == -3
    print("0124 OK")

File: cpp/problems/leetcode/0124_binary_tree_maximum_path_sum/solution.cpp
// LeetCode 124
#include <alg_std.hpp>
#include <cassert>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
    TreeNode(int v, TreeNode* l, TreeNode* r) : val(v), left(l), right(r) {}
};

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        best = INT_MIN;
        dfs(root);
        return best;
    }

private:
    int best;
    int dfs(TreeNode* node) {
        if (!node) return 0;
        int l = max(dfs(node->left), 0);
        int r = max(dfs(node->right), 0);
        best = max(best, node->val + l + r);
        return node->val + max(l, r);
    }
};

int main() {
    TreeNode a(2), b(3), r1(1, &a, &b);
    assert(Solution().maxPathSum(&r1) == 6);
    TreeNode n15(15), n7(7), n20(20, &n15, &n7), n9(9);
    TreeNode r2(-10, &n9, &n20);
    assert(Solution().maxPathSum(&r2) == 42);
    TreeNode r3(-3);
    assert(Solution().maxPathSum(&r3) == -3);
    cout << "0124 OK" << endl;
    return 0;
}
